### PR TITLE
Implement absence endpoints

### PR DESCRIPTION
**Pull request by benj56**

### CODE REVIEW

File: src/absences/index.ts [hunk 1]
```diff
@@ -16,16 +16,33 @@ export default class Absences extends BaseApi {
...
+
+  public cancel(data: AbsenceCancel): Promise<Absence> {
+    const response = this._put<Absence>(`${this.getResourceName()}/cancel`, data);
+    return responseHandler.required(response);
+  }
+
+  public delete(id: number): Promise<Absence> {
+    const response = this._delete<Absence>(`${this.getResourceName()}/delete/${id}`);
+    return responseHandler.required(response);
   }
```
Comment (haubitzer) [inline src/absences/index.ts:47]: validation endpoint is missing

File: src/absences/absences.test.ts [hunk 2]
```diff
@@ -37,7 +40,52 @@ describe('Absences', () => {
...
   test('readById', async () => {
     mock.onGet(`${readPath}/1`).reply(200, { Success: true, NumResults: 1, Results: [{}] });
-    resultSingle = absences.readById(1);
-    await resultSingle.then((result) => expect(result).toStrictEqual({}));
+    const result: Absence = await absences.readById(1);
+    expect(result).toStrictEqual({});
+  });
+
+  test('create', async () => {
+    const absence = { user_id: 1, from_date: '01-01-2020', to_date: '03-01-2020', type_id: 2, subtype_id: 3 };
```
Comment (haubitzer) [inline src/absences/absences.test.ts:48]: Is there any reason that we need to have "real" data here? Especially when the data doesn't make any sense.

File: src/absences/absences.test.ts [hunk 3]
```diff
@@ -37,7 +40,52 @@ describe('Absences', () => {
...
+    const absenceReject: AbsenceReject = { id: 4, granted_comment: 'reasons' };
+    mock.onPut(rejectPath).reply(...successfulResponse(absence));
+    const result: Absence = await absences.reject(absenceReject);
+    expect(result).toStrictEqual(absence);
+  });
+
+  test('cancel', async () => {
+    const absence = { id: 5, user_id: 1, from_date: '01-01-2020', to_date: '03-01-2020', type_id: 2, subtype_id: 3 };
+    const absenceCancel: AbsenceCancel = { id: 5 };
+    mock.onPut(cancelPath).reply(...successfulResponse(absence));
```
Comment (haubitzer) [inline src/absences/absences.test.ts:87]: Check for the correct parameter is missing...

See timetackings tests
mock.onGet(readPath, { params: { id: '99', _op__id: 'eq' } }).reply(

File: src/absences/types.ts [hunk 4]
```diff
@@ -62,3 +62,19 @@ export type Absence = {
...
+
+export interface AbsenceApprove {
+  id: number;
+  granted_comment?: string;
+  substitute_enabled?: boolean;
+}
+
+export interface AbsenceReject extends AbsenceApprove {}
+
+export interface AbsenceCancel {
```
Comment (haubitzer) [inline src/absences/types.ts:74]: When absenceCancel only has an id as a field, we can remove this type and to the same, as we have for delete.. call the endpoint with a number as an argument.

File: test/utils/successfulResponse.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { ApiResponseOnSuccess } from '../../src/utils/response/apiResponse';
```
Comment (haubitzer) [inline test/utils/successfulResponse.ts:1]: I don't like this file very much, I think it hides some information that we want to have visible when writing tests.

**Review by haubitzer — CHANGES_REQUESTED** [review]
Also, update readMe and index.ts

File: src/absences/absences.test.ts [hunk 6]
```diff
@@ -1,32 +1,35 @@
...
     mock.reset();
-    result = null;
-    resultSingle = null;
   });
 
   test('read', async () => {
     mock.onGet(readPath).reply(200, { Success: true, NumResults: 1, Results: [{}] });
-    result = absences.read();
-    await result.then((result) => expect(result).toStrictEqual([{}]));
+    const result: Absence[] = await absences.read();
```
Comment (haubitzer) [inline src/absences/absences.test.ts:26]: What is the reason for this change?

**Review by haubitzer — CHANGES_REQUESTED** [review]

File: src/absences/absences.test.ts (hunk 6 above)
Comment (benj56) [reply src/absences/absences.test.ts:26]: Removing `vars` that need to be reset in `afterEach`. I believe they were added to verify that the type of eg. `absences.read()` is `Promise<Absence[]>`, so I made sure to keep that type verification by explicitly specifying the type on the result variable inside of each test. Then we also don't need different variables like `result` and `resultSingle`. I generally like to replace `then` with `await` to improve readability by removing unnecessary arrow functions.

File: src/absences/index.ts (hunk 1 above)
Comment (benj56) [reply src/absences/index.ts:54]: Added 👍

File: test/utils/successfulResponse.ts (hunk 5 above)
Comment (benj56) [reply test/utils/successfulResponse.ts]: Removed

File: src/absences/absences.test.ts (hunk 2 above)
Comment (benj56) [reply src/absences/absences.test.ts]: Not necessarily need, but I believe it is better than using just `{}` in the tests, where we cannot detect at all if there is some incorrect change or mixing up of fields somewhere between receiving the response and returning from eg. `timeTrackings.read()`. Not necessary at the moment since there is no logic there, but could become relevant if some sort of transformation needs to be done. With "doesn't make any sense", were you referring to the previously incorrect date format?

File: src/absences/absences.test.ts (hunk 3 above)
Comment (benj56) [reply src/absences/absences.test.ts]: Ah, nice, added it. Did not realize this was possible, it only seems to be used in one timeTrackings test. Should we make plans to add this to all tests containing a request with params or body? 

Sadly this only gives 404 if the request does not match, I experimented with adding an assertion to the reply callback instead, so that we get the nice assertion error message containing the diff. In the end I removed it again as it is probably too much code for this small benefit but you can take a look here if you are interested: https://github.com/TimeTac/js-client-library/pull/30/commits/90edfb48486685ecddb26a5cee8074d7b77943ba

File: src/utils/response/responseHandlers.ts [hunk 7]
```diff
@@ -1,58 +1,53 @@
 import { AxiosResponse } from 'axios';
-import { ApiResponse, ApiResponseOnSuccess } from './apiResponse';
+import { ApiResponse } from './apiResponse';
 
-class ResponseHandler {
-  async toApiResponse<T>(promise: Promise<AxiosResponse>) {
-    const dataPromise = await promise;
-    const apiResponse = (dataPromise.data as unknown) as ApiResponse<T>;
```
Comment (benj56) [inline src/utils/response/responseHandlers.ts:7]: Now that we have correct types coming in, I was able to remove this `as unknown as ApiResponse` cast

File: src/absences/types.ts (hunk 4 above)
Comment (benj56) [reply src/absences/types.ts]: done

File: src/baseApi.ts [hunk 8]
```diff
@@ -27,19 +27,19 @@ export default abstract class BaseApi {
     };
   }
 
-  protected _get<T>(endpoint: string, options?: AxiosRequestConfig): Promise<AxiosResponse> {
+  protected _get<T>(endpoint: string, options?: AxiosRequestConfig) {
```
Comment (AlGvrl) [inline src/baseApi.ts:30]: Why remove the return types?

File: src/utils/response/responseHandlers.ts [hunk 9]
```diff
@@ -1,58 +1,53 @@
...
+/**
+ * @return A promise that resolves to T or undefined if no results but Success is true.
+ */
+export async function optional<T>(promise: RequestPromise<T[]>): Promise<T | undefined> {
+  const response = await toApiResponse<T[]>(promise);
+  return (response.NumResults > 0 && response.Results[0]) || undefined;
+}
+
+/**
+ * @return A promise that resolves to array of T or null if no results but Success is true.
```
Comment (AlGvrl) [inline src/utils/response/responseHandlers.ts:48]: Out of scope in this PR, but having this resolve to "array of T or null", while `optional()` resolves to "T or undefined" seems like something that might come back and bite us in one place in the future.

Also, unless I'm missing something, it's not even an "array of null" -- just an empty array.

File: src/absenceDays/index.ts [hunk 10]
```diff
@@ -1,20 +1,20 @@
...
-    return responseHandler.requiredList(response);
+    return requiredList(response);
   }
 
   public readById(id: number, requestParams?: RequestParams<AbsenceDay> | Object): Promise<AbsenceDay> {
     const params = requestParams instanceof RequestParams ? requestParams.getParams() : requestParams;
-    const response = this._get<AbsenceDay>(`${this.getResourceName()}/read/${id}`, { params });
-    return responseHandler.required(response);
+    const response = this._get<AbsenceDay[]>(`${this.getResourceName()}/read/${id}`, { params });
+    return required(response);
```
Comment (AlGvrl) [inline src/absenceDays/index.ts:18]: I think I personally would've preferred the old style, with `responseHandler.required()`, since it's more explicit and having just `required()` confuses me a bit, but it's a matter of taste, I guess.

Comment (benj56) [reply src/absenceDays/index.ts:18]: I would also prefer that. I removed it since responseHandler was a class and needed to be instantiated before exporting. Do you think exporting it like the following, so that we can still call it with `responseHandler.something()` would make sense instead? 
`export const responseHandlers = { required, requiredList... }`

File: src/baseApi.ts (hunk 8 above)
Comment (benj56) [reply src/baseApi.ts:30]: The actual type returned and now inferred is `Promise<AxiosResponse<ApiResponse<T>>>` I believe. I deleted the explicit type as it removed some of the information.

File: src/utils/response/responseHandlers.ts (hunk 9 above)
Comment (benj56) [reply src/utils/response/responseHandlers.ts:48]: yeah, that comment is wrong, this just returns an array, never null or undefined. Not sure why we even check for NumResults, we should be able to just return response.Results as-is. Also I'm not sure whether requiredList is even needed and I have the feeling that it might be in use in some places where optionalList should be used. Ideally we can remove requiredList and have optionalList be just "list".  Probably would be best to create a ticket for that

File: src/absenceDays/index.ts (hunk 10 above)
Comment (AlGvrl) [reply src/absenceDays/index.ts:18]: Possibly, but I'm not sure if changes to the export are necessary, you can do something like
```
import * as responseHandlers from '../utils/response/responseHandlers';
```
(nor sure if i got the syntax exactly right though)

File: src/utils/response/responseHandlers.ts (hunk 7 above)
Comment (haubitzer) [reply src/utils/response/responseHandlers.ts:7]: @benj56 Please keep different changes separated in the PR. It is very hard to see why these changes were needed in the context of the story. If you like to improve something just create a new story and don't just implement it in the current branch.

**Review by haubitzer — CHANGES_REQUESTED** [review]

File: src/absences/absences.test.ts [hunk 11]
```diff
@@ -37,7 +40,58 @@ describe('Absences', () => {
...
   test('readById', async () => {
     mock.onGet(`${readPath}/1`).reply(200, { Success: true, NumResults: 1, Results: [{}] });
-    resultSingle = absences.readById(1);
-    await resultSingle.then((result) => expect(result).toStrictEqual({}));
+    const result: Absence = await absences.readById(1);
+    expect(result).toStrictEqual({});
+  });
+
+  test('create', async () => {
+    const absenceCreate: AbsenceCreate = { user_id: 1, from_date: '2020-01-01', to_date: '2020-01-03', type_id: 2, subtype_id: 3 };
```
[the remark this answers is not in the document]
Comment (benj56) [reply src/absences/absences.test.ts:48]: As said, the benefit would be detecting mistakes in the endpoint file, if we had any transformations there. For example, change the returned absence in create() and replace user_id with id. Then the test with "real data" fails and shows the diff, while the one with empty object would pass anyway.

However, I agree that it's not necessary for now as I don't foresee adding any transformations anyway. So, we could also remove this "real" data again. Let me know what you would prefer

[the remark this answers is not in the document]
Comment (haubitzer) [reply src/absences/absences.test.ts:48]: I would remove it because we had to maintain the "real data". Integration tests will ensure that the data types are correct. And when we have transformations we can test them with an own set of unit tests.

[the remark this answers is not in the document]
Comment (benj56) [reply src/absences/absences.test.ts:48]: Removed it

**Review by haubitzer — APPROVED** [review]

**Review by AlGvrl — APPROVED** [review]
